gt/main.c

```diff
@@ -244,13 +250,94 @@ print_ip_err_msg(struct gt_packet_headers *pkt_info)
 		"gt: receiving a packet with IP source address %s, and destination address %s, whose destination IP address is not the Grantor server itself.!\n",
 		src, dst);
 }
+static void
+gt_arp_and_nd_req_cb(const struct lls_map *map, void *arg,
+	__attribute__((unused))enum lls_reply_ty ty, int *pcall_again)
+{
+	struct ether_cache *eth_cache = arg;
+
+	ether_addr_copy(&map->ha, &eth_cache->eth_hdr.d_addr);
+	eth_cache->stale = map->stale;
+
+	if (pcall_again != NULL)
+		*pcall_again = true;
+}
+
+/*
+ * Fill up the Ethernet cached header.
+ * Note that the destination MAC address should be filled up by LLS.
+ * The fields @ref_cnt and @ip_addr are not used by GT blocks.
+ */
+static inline int
+gt_fill_up_ether_cache_locked(struct ether_cache *eth_cache,
+	uint16_t inner_ip_ver, void *ip_dst,
+	struct gatekeeper_if *iface, int lcore_id)
+{
+	int ret;
+
+	eth_cache->stale = true;
+	eth_cache->eth_hdr.ether_type = rte_cpu_to_be_16(inner_ip_ver);
+	ether_addr_copy(&iface->eth_addr, &eth_cache->eth_hdr.s_addr);
+
+	if (inner_ip_ver == ETHER_TYPE_IPv4) {
+		ret = hold_arp(gt_arp_and_nd_req_cb,
+			eth_cache, ip_dst, lcore_id);
+		if (ret < 0)
+			return ret;
+	} else if (likely(inner_ip_ver == ETHER_TYPE_IPv6)) {
+		ret = hold_nd(gt_arp_and_nd_req_cb,
+			eth_cache, ip_dst, lcore_id);
+		if (ret < 0)
+			return ret;
+	} else
+		return -1;
+
+	return 0;
+}
+
+static struct ether_cache *
+gt_neigh_get_ether_cache(struct neighbor_hash_table *neigh,
+	uint16_t inner_ip_ver, void *ip_dst,
+	struct gatekeeper_if *iface, int lcore_id)
+{
+	int ret;
+	struct ether_cache *eth_cache = lookup_ether_cache(neigh, ip_dst);
+	if (eth_cache != NULL)
+		return eth_cache;
+
+	ret = rte_hash_add_key(neigh->hash_table, ip_dst);
+	if (ret >= 0) {
+		int succ;
+		eth_cache = &neigh->cache_tbl[ret];
+		succ = gt_fill_up_ether_cache_locked(
+			eth_cache, inner_ip_ver, ip_dst, iface, lcore_id);
+		if (succ < 0) {
+			RTE_VERIFY(rte_hash_del_key(
+				neigh->hash_table, ip_dst) == ret);
+			initialize_ether_cache(eth_cache);
```

Comment:
If the GK integration with LLS is merged first, I think this will change to `clear_ether_cache()`.